### ISSUE TITLE
Discussion: Developer/Strongloop communication

### ISSUE THREAD

**Issue by doublemarked** [strongloop/loopback#3042]
At the advice of @bajtos in #2693 I am opening this discussion regarding developer communication.

Context: Fixes #2693 and #3021 implemented access-token invalidation when the users changes an email or a password. However, the current code deletes all access tokens, including the token used to make the change password/email request. The effect of this is that client code must anticipate immediate token invalidation any time it submits changes to a user's email and password, which is notable behavior change to be slipped into a point release.

I'd like to propose several questions associated with this. These questions are fielded to both the Strongloop team and the developer community,

1. Under what conditions do we consider it acceptable to introduce breaking changes into a minor release?
2. What is the best way in which the Strongloop team can communicate to the developer community when such changes are upcoming?

### Tasks

 - [x] Setup a low-volume Announcement mailing list - see https://groups.google.com/forum/#!forum/loopbackjs-announcements
 - [x] Update loopback.io documentation - https://github.com/strongloop/loopback.io/pull/316
 - [x] Update [README](https://github.com/strongloop/loopback#resources) in this repository - see https://github.com/strongloop/loopback/pull/3261

**Comment by doublemarked** [comment]
I'd like to share my own thoughts on this. I may have missed it, but I don't recall any communication regarding these changes, which I think took place in mid-September. Our team did not discover the problem until last week when we performed some cleanup that bumped up the Loopback version. 

I would like to offer the following suggestions, 

  1. Perhaps Strongloop should consider a low-volume announcement-only Google Group that is used for important developer alerts/notifications (security notices, breaking changes, notable releases, etc). The general Google Group has a lot of chatter and I would not be surprised if many developers do not subscribe or ignore it. 

  2. Similarly, perhaps the Github release notes for 2.35.0 should have emphasized the breaking changes. 
 
  3. Might it have been been more appropriate to include these changes only with configuration that allows it to be disabled, even if it defaults to on?

**Comment by superkhau** [comment]
@doublemarked Thanks for bringing up this discussion. I've been advocating a Community Governance page on loopback.io for things like this (and transparency especially). @bajtos Has something in the backlog (probably won't get to it till the new year though since most people are going on vacation soon). 

IMO:

>Under what conditions do we consider it acceptable to introduce breaking changes into a minor release?
Never since we should be following semver and only breaking on majors.

>What is the best way in which the Strongloop team can communicate to the developer community when such changes are upcoming?
IMO, Google Groups needs to be first and foremost our annoucement channel (in addition to blog posts/tweets). I brought up some discussions in the past and loopback-announce was debated (no action taken yet).

>Similarly, perhaps the Github release notes for 2.35.0 should have emphasized the breaking changes.
I didn't follow the discussion above, but IMO it should've never broken in the first place. That said, I think the better question is what do we do when accidentally releasing breaking changes in minor/patch versions. Do you have any suggestions for this? Maybe doing another subsequent minor/patch to revert the change as soon as reported?

>Might it have been been more appropriate to include these changes only with configuration that allows it to be disabled, even if it defaults to on?

The policy should be to default to the state it was previously at to prevent breaking changes -- this is not officially documented, but I propose we set this as a ground rule.

**Comment by doublemarked** [comment]
> Never since we should be following semver and only breaking on majors.

Yeah, indeed. I have taken a cautious tone because it's not clear whether anybody agrees with me that token invalidation is a breaking feature, but I certainly agree with following semver. 

> I didn't follow the discussion above, but IMO it should've never broken in the first place. That said, I think the better question is what do we do when accidentally releasing breaking changes in minor/patch versions. Do you have any suggestions for this? Maybe doing another subsequent minor/patch to revert the change as soon as reported?

Yes I think a patch that reverts the change (or introduces a default-off flag) would be correct. It should be treated with similar severity to a regression. 

> The policy should be to default to the state it was previously at to prevent breaking changes (and force users to manually turn on the feature flag) -- this is not officially documented, but I propose we set this as a ground rule.

Sounds good to me.

**Comment by bajtos** [comment]
Hi, thank you @doublemarked for starting this discussion.

> Under what conditions do we consider it acceptable to introduce breaking changes into a minor release?

I agree with what @superkhau wrote, we shouldn't be introducing breaking changes into minor versions.

In the particular case you have described (access-token invalidation), we did not realise this change may be breaking existing applications :( To be honest, this seems like a problem that cannot be fixed - how to know what we don't know? The best we can do is to limit the impact of changes like this one.

> Yes I think a patch that reverts the change (or introduces a default-off flag) would be correct. It should be treated with similar severity to a regression.

I like the idea of using feature flags and changing their default value to quickly enable/disable the feature if needed 👍 

>  Perhaps Strongloop should consider a low-volume announcement-only Google Group that is used for important developer alerts/notifications (security notices, breaking changes, notable releases, etc). The general Google Group has a lot of chatter and I would not be surprised if many developers do not subscribe or ignore it.

+1 ✖️ 💯 from me.

**Comment by superkhau** [comment]
tldr;

- [ ] Create loopback-announce for low-volume announcements
- [ ] Feature flags to enable/disable features (breaking changes) that were accidentally introduced in non-major versions
- [ ] Community governance page to describe such processes for transparency

We'll need to create issues to take these items on in the new year (or whenever we decide to prioritize these issues as part of sprint planning).

**Comment by superkhau** [comment]
Slapping 8 point estimates to @ritch to prioritize during backlog grooming.

**Comment by doublemarked** [comment]
All sounds great, guys. TYVM!

**Comment by crandmck** [comment]
Please also open a PR in loopback.io to add information about for the new GG and feature flags to the loopback.io site.  If unsure where or how to do so, open an issue in the loopback.io repo and I will help.

**Comment by superkhau** [comment]
@crandmck Updated bullet list to include tasks from your comments. TY to adding in.

**Comment by bajtos** [comment]
Let's work in smaller incremental steps here. I created https://github.com/strongloop/loopback/issues/3257 to track the documentation about feature flags.

I'd like to keep the scope of this item only as the Announcements Google Group together with necessary loopback.io changes.

**Comment by bajtos** [comment]
The new mailing list: https://groups.google.com/forum/#!forum/loopbackjs-announcements

**Comment by doublemarked** [comment]
Excellent!

**Comment by bajtos** [comment]
I configured the Google Group permissions to allow anybody to subscribe to this mailing list, but also to restrict new posts to administrators only. We can tweak these settings in the future as needed.

**Comment by bajtos** [comment]
Announcement post in LoopBackJS group: https://groups.google.com/forum/#!topic/loopbackjs/qvMC6r2IRwo

**Comment by bajtos** [comment]
Done.
